azure/services/bastionhosts/azurebastion.go

```diff
@@ -0,0 +1,225 @@
+	publicIP, err := s.publicIPsClient.Get(ctx, s.Scope.ResourceGroup(), azureBastionSpec.PublicIPName)
+	if err != nil && azure.ResourceNotFound(err) {
+		publicIP, err = s.createBastionPublicIP(ctx, azureBastionSpec.PublicIPName)
+		if err != nil {
+			return errors.Wrap(err, "failed to create bastion publicIP")
+		}
+	} else if err != nil {
+		return errors.Wrap(err, "failed to get existing publicIP")
+	}
+	s.Scope.V(2).Info("successfully got public ip", "publicIP", azureBastionSpec.PublicIPName)
+
+	s.Scope.V(2).Info("checking if subnet exists otherwise will try to create", "subnet", azureBastionSpec.SubnetSpec)
+	subnet, err := s.subnetsClient.Get(ctx, s.Scope.ResourceGroup(), azureBastionSpec.VNetName, azureBastionSpec.SubnetSpec.Name)
+	if err != nil && azure.ResourceNotFound(err) {
+		subnet, err = s.createSubnet(ctx, azureBastionSpec.SubnetSpec)
+		if err != nil {
+			return errors.Wrap(err, "failed to create subnet")
+		}
+	} else if err != nil {
+		return errors.Wrap(err, "failed to get existing publicIP")
+	}
+	s.Scope.V(2).Info("successfully got subnet", "subnet", azureBastionSpec.SubnetSpec)
+
+	s.Scope.V(2).Info("creating bastion host", "bastion", azureBastionSpec.Name)
+	bastionHostIPConfigName := fmt.Sprintf("%s-%s", azureBastionSpec.Name, "bastionIP")
+	err = s.client.CreateOrUpdate(
+		ctx,
+		s.Scope.ResourceGroup(),
+		azureBastionSpec.Name,
+		network.BastionHost{
+			Name:     to.StringPtr(azureBastionSpec.Name),
+			Location: to.StringPtr(s.Scope.Location()),
+			Tags: converters.TagsToMap(infrav1.Build(infrav1.BuildParams{
+				ClusterName: s.Scope.ClusterName(),
+				Lifecycle:   infrav1.ResourceLifecycleOwned,
+				Name:        to.StringPtr(azureBastionSpec.Name),
+				Role:        to.StringPtr("Bastion"),
+			})),
+			BastionHostPropertiesFormat: &network.BastionHostPropertiesFormat{
+				DNSName: to.StringPtr(fmt.Sprintf("%s-bastion", strings.ToLower(azureBastionSpec.Name))),
+				IPConfigurations: &[]network.BastionHostIPConfiguration{
+					{
+						Name: to.StringPtr(bastionHostIPConfigName),
+						BastionHostIPConfigurationPropertiesFormat: &network.BastionHostIPConfigurationPropertiesFormat{
+							Subnet: &network.SubResource{
+								ID: subnet.ID,
+							},
+							PublicIPAddress: &network.SubResource{
+								ID: publicIP.ID,
+							},
+							PrivateIPAllocationMethod: network.Dynamic,
+						},
+					},
+				},
+			},
+		},
+	)
+	if err != nil {
+		return errors.Wrap(err, "cannot create Azure Bastion")
+	}
+
+	s.Scope.V(2).Info("successfully created bastion host", "bastion", azureBastionSpec.Name)
+	return nil
+}
+
+func (s *Service) ensureAzureBastionDeleted(ctx context.Context, azureBastionSpec azure.AzureBastionSpec) error {
+	s.Scope.V(2).Info("deleting bastion host", "bastion", azureBastionSpec.Name)
+
+	err := s.client.Delete(ctx, s.Scope.ResourceGroup(), azureBastionSpec.Name)
+	if err != nil && azure.ResourceNotFound(err) {
+		// already deleted
+		return nil
```

Comment:
I don't think we can return here since we'll need to check for other resources

Reply:
Good catch, thanks